### PR TITLE
Add binary_cross_entropy and trace decomp - fixed _log_softmax/_softmax dtype promotion semantics

### PR DESCRIPTION
**Pull request by Chillee**
cc: @zou3519

### CODE REVIEW

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -460,6 +460,26 @@ def nll_loss_backward(
...
+    self: Tensor,
+    target: Tensor,
+    weight: Optional[Tensor] = None,
+    reduction: int = Reduction.MEAN.value,
+) -> Tensor:
+    # We cannot currently model this without introducing data-dependent control flow
+    # TORCH_CHECK(
+    #     (input_val >= 0) && (input_val <= 1),
+    #     "all elements of input should be between 0 and 1"
+    # )
```
Comment (zou3519) [inline torch/_decomp/decompositions.py:474]: Nit: It's not clear which of {self, target, weight} is "input_val", but I guess it doesn't matter much until we need to figure out how to model this

**Comment by zou3519** [comment]
CI failures seem real?

**Comment by Chillee** [comment]
CI failures look like rebase issues I think.

**Review by ezyang — APPROVED** [review]
I wish there was an easy way to tell that test coverage automatically increased to cover thsi lol

**Pull request closed without merging**